packages/@internationalized/number/src/NumberParser.ts

```diff
@@ -180,34 +200,108 @@ class NumberParserImpl {
+          integerPart = (
+            parseInt(
+              integerPart
+              .replace(NUMERALS_ARAB_REGEX, (d) => NUMERALS_ARAB.indexOf(d).toString()),
+              10
+            ).toString()
+            .replace(NUMERALS_REGEX, (d) => NUMERALS_ARAB.split('')[parseInt(d, 10)])
+          );
+        } else if (isHanidec) {
+          // Replace Hanidec numerals with Latin numerals,
+          // then parse the integer part to remove leading zeros,
+          // and finally replace Latin numerals with Hanidec numerals.
+          integerPart = (
+            parseInt(
+              integerPart
+              .replace(NUMERALS_HANIDEC_REGEX, (d) => NUMERALS_HANIDEC.indexOf(d).toString()),
+              10
+            ).toString()
+            .replace(NUMERALS_REGEX, (d) => NUMERALS_HANIDEC.split('')[parseInt(d, 10)])
+          );
+        } else {
+          integerPart = parseInt(integerPart, 10).toString();
+        }
+        let decimalSymbol = decimalSymbolMatch.length > 1 && lastDecimalSymbol === firstDecimalSymbol ? '' : lastDecimalSymbol;
+        if (decimalSymbol !== '') {
+          if (this.symbols.decimal &&
+            lastDecimalSymbol !== this.symbols.decimal &&
+            !isArabic &&
+            !isHanidec &&
+            (
+              integerPart.length > 3 ||
+              integerPart === '0' ||
+              (firstDecimalSymbol === this.symbols.decimal && lastDecimalSymbol === this.symbols.group) ||
+              (decimalSymbolMatch.length === 1 && decimalPart.length > 3)
+            )
+          ) {
+            decimalSymbol = this.symbols.decimal;
+          }
+        }
+
+        abs = `${integerPart ?? ''}${decimalSymbol}${decimalPart ?? ''}`;
+      } else if (groupSymbolMatch) {
+        abs = parseInt(abs.replace(GROUPING_SYMBOLS_REGEX, ''), 10).toString();
+      }
+      sanitizedValue = `${beforeAbs}${abs}${afterAbs}`;
+    }
+
     // Remove literals and whitespace, which are allowed anywhere in the string
-    value = value.replace(this.symbols.literals, '');
+    sanitizedValue = sanitizedValue.replace(this.symbols.literals, '');
 
     // Replace the ASCII minus sign with the minus sign used in the current locale
     // so that both are allowed in case the user's keyboard doesn't have the locale's minus sign.
     if (this.symbols.minusSign) {
-      value = value.replace('-', this.symbols.minusSign);
+      sanitizedValue = sanitizedValue.replace(MINUS_SIGN_REGEX, this.symbols.minusSign);
     }
 
     // In arab numeral system, their decimal character is 1643, but most keyboards don't type that
     // instead they use the , (44) character or apparently the (1548) character.
     if (this.options.numberingSystem === 'arab') {
       if (this.symbols.decimal) {
-        value = value.replace(',', this.symbols.decimal);
-        value = value.replace(String.fromCharCode(1548), this.symbols.decimal);
+        sanitizedValue = sanitizedValue.replace(',', this.symbols.decimal);
```

Comment:
do we still need some of these replacement lines with the sanitize changes you've made?